<launch>
  
  <!-- Arguments for EKF -->
  <arg name="sampling_period" default="0.01" /> 
  <arg name="system_noise_p" default="0.000" /> 
  <arg name="system_noise_v" default="0.001" /> 
  <arg name="system_noise_th" default="0.00" /> 
  <arg name="system_noise_ba" default="0.01" /> 
  <arg name="measurement_noise_p" default="0.02" /> 
  <arg name="measurement_noise_v" default="0.01" /> 
  <arg name="measurement_noise_th" default="20.2" /> 
  
  <!-- Arguments for Kinect driver -->
  <arg name="recording_file" default="" />
  <arg name="recording_loop_enabled" default="true" />
  <arg name="belt_speed" default="0.0" /> 

  <!-- Arguments for sport_sole_publisher -->
  <arg name="watch_sport_sole" default="true" />

  <!-- Arguments for goal_generator -->
  <arg name="max_num_laps" default="3" />

  <!-- Arguments for straight line walking -->
  <arg name="straight_open_loop_enabled" default="false" />
  <arg name="straight_distance" default="5.0" />
  <arg name="v_max" default="1.2" />

  <!-- Arguments for rosbag record -->
  <arg name="record_bag" default="true" />
  <arg name="play_bag" default="false" />
  <arg name="bag_name" default="data" />

  <!-- Arguments for gait_analyzer -->
  <arg name="enable_gait_analyzer" default="false" />
  <arg name="gait_analyzer_global_frame" default="odom" />
  <arg name="record_gait_analytics" default="true" />

  <!-- Other arguments -->
  <!-- static_test: is the robot static? -->
  <arg name="static_test" default="false" />
  <arg name="waypoints_file_suffix" default="_cw"/>
  <arg name="disable_rviz" default="false"/>
  <arg name="database_path" default="rtabmap.db"/>

  <!-- <node name="gait_analyzer" pkg="gait_training_robot" type="gait_analyzer" output="screen" >
    <param name="belt_speed" value="$(arg belt_speed)" />
  </node> -->

  <param name="robot_description2" textfile="$(find gait_training_robot)/urdf/gtr2.urdf" />
  <group unless="$(arg disable_rviz)">
    <node name="rviz" pkg="rviz" type="rviz" args="-d $(find gait_training_robot)/rviz/slam_rtabmap_ekf.rviz" output="screen" unless="$(arg enable_gait_analyzer)" />
  </group>

  <!-- ++++ FILTER INPUTS ++++ -->
  <!-- Use real data -->
  <group unless="$(arg play_bag)">

    <!-- sport_sole_publisher -->
    <node pkg="sport_sole" type="sport_sole_publisher" name="sport_sole_publisher" output="screen" />
    <!-- <node pkg="rviz" type="rviz" name="rviz" args="-d $(find sport_sole)/rviz/accel_orient.rviz" if="$(arg watch_sport_sole)"/> -->

    <!-- If (static_test==true) -->
    <group if="$(arg static_test)">
      <!-- Kinect driver launch file -->
      <include file="$(find azure_kinect_ros_driver)/launch/driver.launch" >
        <arg name="depth_mode" value="WFOV_2X2BINNED" />
        <arg name="body_tracking_enabled" value="true" />
        <arg name="fps" value="30" />
        <arg name="rgb_point_cloud" value="false" />
        <arg name="recording_file" value="$(arg recording_file)" />
        <arg name="recording_loop_enabled" value="$(arg recording_loop_enabled)" />
      </include>
      <!-- Broadcast tf from map to camera_link -->
      <node name="dummy_broadcaster_static" pkg="tf" type="static_transform_publisher" 
            args="0 0 0.70 0 0 1 0 map camera_base 100" if="$(arg static_test)" />
    </group>

    <!-- Else if (static_test==false), start the robot -->
    <group unless="$(arg static_test)">
      <include file="$(find gait_training_robot)/launch/test1_rtabmap_headless.launch" >
        <arg name="localization" value="true" />
        <arg name="enable_distance_controller" value="true"/>
        <arg name="open_loop_controller_enabled" value="$(arg straight_open_loop_enabled)"/>
        <arg name="open_loop_controller_terminal_distance" value="$(arg straight_distance)"/>
        <arg name="open_loop_controller_wait_for_required_msgs" value="true"/>
        <arg name="azure_kinect_driver_required" default="true" />
        <arg name="recording_file" value="$(arg recording_file)" />
        <arg name="recording_loop_enabled" value="$(arg recording_loop_enabled)" />
        <arg name="enable_goal_generator" value="true" />
        <arg name="max_num_laps" value="$(arg max_num_laps)" />
        <arg name="delay_start_secs" value="20.0" />
        <arg name="v_max" value="$(arg v_max)" />
        <arg name="suffix" value="$(arg waypoints_file_suffix)"/>
        <arg name="wired_sync_enabled" value="true" />
        <param name="database_path" type="string" value="$(arg database_path)"/>
      </include>
    </group>

  </group>

  <!-- rosbag play -->
  <group if="$(arg play_bag)">
    <param name="use_sim_time" value="true"/>
    <node name="rosbag_play_node" pkg="rosbag" type="play" required="true"
          args="--clock $(find gait_training_robot)/bags/$(arg bag_name).bag"
          if="$(arg play_bag)" />
    <!-- Robot state publisher -->
    <param name="robot_description" textfile="$(find gait_training_robot)/urdf/gtr2.urdf" />
	  <node name="robot_state_publisher_gtr2" pkg="robot_state_publisher" type="robot_state_publisher" unless="$(arg disable_rviz)"/>
  </group>

  <!-- rosbag record -->
  <group if="$(arg static_test)">
    <node name="rosbag_record_node" pkg="rosbag" type="record" 
          args="record -o $(find gait_training_robot)/bags/data 
                /sport_sole_publisher/sport_sole
                /body_tracking_data /tf /tf_static" 
          if="$(arg record_bag)"/>
  </group>

  <group unless="$(arg static_test)">
    <node name="rosbag_record_node" pkg="rosbag" type="record" 
          args="record -o $(find gait_training_robot)/bags/$(arg bag_name)
                /sport_sole_publisher/sport_sole
                /sport_sole_publisher/t0_zeno
                /imu /odom /cmd_vel
                /body_tracking_data /tf /tf_static
                /joint_states"
          if="$(arg record_bag)"/> 
    
  </group>

  <!-- gait_analyzer -->
  <group if="$(arg enable_gait_analyzer)" >
    <node name="gait_analyzer" pkg="gait_training_robot" type="gait_analyzer" output="screen" >
      <param name="belt_speed" value="$(arg belt_speed)" />
      <param name="global_frame" value="$(arg gait_analyzer_global_frame)" />
    </node>
    <node name="rviz" pkg="rviz" type="rviz" args="-d $(find gait_training_robot)/rviz/slam_rtabmap_ekf_ga.rviz" />
    <node name="rosbag_record_ga_node" pkg="rosbag" type="record" 
          args="record -o $(find gait_training_robot)/bags/$(arg bag_name)_ga 
                --regex '(/gait_analyzer/.*|/kinect_pose_estimator/.*)'" 
          if="$(arg record_gait_analytics)"/> 
  </group>

</launch>